onnx/inliner/inliner.cc

```diff
@@ -246,103 +200,114 @@ class Specializer {
   }
 
   // Process a node:
-  void Transform(NodeProto& n) {
-    if (!n.name().empty())
-      n.set_name(MakeUnique(n.name()));
+  bool ProcessNode(NodeProto* node) override {
+    if (!node->name().empty())
+      node->set_name(MakeUnique(node->name()));
 
-    for (auto& x : *n.mutable_input()) {
+    for (auto& x : *node->mutable_input()) {
       LookupOrRename(x, false);
     }
-    for (auto& y : *n.mutable_output()) {
+    for (auto& y : *node->mutable_output()) {
       LookupOrRename(y, true);
     }
-    auto& attributes = *n.mutable_attribute();
-    for (auto attr_iter = attributes.begin(); attr_iter != attributes.end();) {
-      auto& attr = *attr_iter;
-      if (!attr.ref_attr_name().empty()) {
-        // Attribute-references must be replaced by the corresponding attribute-value in the call-node
-        // if the call-node contains the attribute. Otherwise, this attribute must be removed.
-        auto* attr_val = attr_map(attr.ref_attr_name());
-        if (attr_val != nullptr) {
-          // Copy value of attribute, but retain original name:
-          std::string name = attr.name();
-          attr = *attr_val;
-          attr.set_name(name);
-        } else {
-          attr_iter = attributes.erase(attr_iter);
-          continue;
-        }
-      }
-      // Subgraphs must be recursively processed.
-      if (attr.has_g()) {
-        Transform(*attr.mutable_g());
-      }
-      for (auto& graph : *attr.mutable_graphs())
-        Transform(graph);
-      ++attr_iter;
-    }
+    return true; // Process attribute subgraphs in traversal
   }
 
   // Process a sub-graph, contained as an attribute in a control-flow op node.
-  void Transform(GraphProto& graph) {
+  // Since we need both pre-processing and post-processing in the traversal, we
+  // override the VisitGraph method.
+  void VisitGraph(GraphProto* graph) override {
```

Comment:
Just for my learning: so graph needs to be a pointer and not a smart pointer or something?

Reply:
Yes.

Reply:
Curious what the reason was?